api/src/main/java/net/kyori/adventure/text/event/ClickCallback.java

```diff
@@ -0,0 +1,259 @@
+   * @return a filtered callback action
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> filter(final @NotNull Predicate<T> filter, final @Nullable Consumer<? super Audience> otherwise) {
+    return audience -> {
+      if (filter.test(audience)) {
+        this.accept(audience);
+      } else if (otherwise != null) {
+        otherwise.accept(audience);
+      }
+    };
+  }
+
+  /**
+   * Require that audiences receiving this callback have a certain permission.
+   *
+   * <p>For audiences without permissions information, this test will always fail.</p>
+   *
+   * <p>Actions from audiences that do not match this predicate will be silently ignored.</p>
+   *
+   * @param permission the permission to check
+   * @return a modified callback
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> requiringPermission(final @NotNull String permission) {
+    return this.requiringPermission(permission, null);
+  }
+
+  /**
+   * Require that audiences receiving this callback have a certain permission.
+   *
+   * <p>For audiences without permissions information, this test will always fail.</p>
+   *
+   * @param permission the permission to check
+   * @param otherwise the action to perform on the audience if the conditions are not met
+   * @return a modified callback
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> requiringPermission(final @NotNull String permission, final @Nullable Consumer<? super Audience> otherwise) {
+    return this.filter(audience -> audience.getOrDefault(PermissionChecker.POINTER, ClickCallbackInternals.ALWAYS_FALSE).test(permission), otherwise);
+  }
+
+  /**
+   * Options to configure how a callback can be executed.
+   *
+   * @since 4.13.0
+   */
+  @ApiStatus.NonExtendable
+  interface Options extends Examinable {
+    /**
+     * Indicate that a callback should have unlimited uses.
+     *
+     * @since 4.13.0
+     */
+    int UNLIMITED_USES = -1;
+
+    /**
+     * Create a new builder.
+     *
+     * @return the new builder
+     * @since 4.13.0
+     */
+    static @NotNull Builder builder() {
+      return new ClickCallbackOptionsImpl.BuilderImpl();
+    }
+
+    /**
+     * Create a new builder populating from existing options.
+     *
+     * @param existing the existing options to populate this builder with
+     * @return the new builder
+     * @since 4.13.0
+     */
+    static @NotNull Builder builder(final @NotNull Options existing) {
+      return new ClickCallbackOptionsImpl.BuilderImpl(existing);
+    }
+
+    /**
+     * The number of times this callback can be executed.
+     *
+     * <p>By default callbacks are single-use.</p>
+     *
+     * @return allowable use count, or {@link #UNLIMITED_USES}
+     * @since 4.13.0
+     */
+    int uses();
+
+    /**
+     * How long this callback will last until it is made invalid.
+     *
+     * <p>By default callbacks last {@link #DEFAULT_LIFETIME 12 hours}.</p>
```

Comment:
id prefer we just have the "12 hours" in the default lifetime annotation so we don't inevitably forget to change this comment